Fix and refine core function escaping in block output

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Reverts various blocks as per audit in https://github.com/WordPress/gutenberg/issues/38568

Closes https://github.com/WordPress/gutenberg/issues/38568

## Testing Instructions

Test each block's output on the front of the site. No raw HTML should be output.


## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

getdave committed:
Comments Date block - revert core function escaping

getdave committed:
Latest comments: revert escaping of core functions

getdave committed:
Latest Posts - revert escaping of core functions

getdave committed:
Page list block - revert unnecessary escaping

getdave committed:
Comments link block - revert escape of post title

getdave committed:
Post date block - remove escape of Core function

getdave committed:
Site tagline block - remove escape of core function

getdave committed:
Site title block - revert escape of core function

getdave commented on packages/block-library/src/comment-date/index.php:
```diff
@@ -31,13 +31,13 @@ function render_block_core_comment_date( $attributes, $content, $block ) {
 	$link               = get_comment_link( $comment );
 
 	if ( ! empty( $attributes['isLink'] ) ) {
-		$formatted_date = sprintf( '<a href="%1s">%2s</a>', esc_url( $link ), esc_html( $formatted_date ) );
```
Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/comment-date/index.php:
```diff
@@ -31,13 +31,13 @@ function render_block_core_comment_date( $attributes, $content, $block ) {
 		'<div %1$s><time datetime="%2$s">%3$s</time></div>',
 		$wrapper_attributes,
-		esc_attr( get_comment_date( 'c', $comment ) ),
+		get_comment_date( 'c', $comment ),
```
Core function - can be trusted, maybe filtered.

Mamaduka replied:
I think it's okay to escape this value - https://github.com/Automattic/_s/blob/master/inc/template-tags.php#L23

getdave replied:
Unfortunately as I'm learning, anything that can be filtered can contain HTML. 

https://github.com/WordPress/wordpress-develop/blob/f0180f59ea00683647c602941c1f2b422192df46/src/wp-includes/comment-template.php#L568

So I'm not sure. Most of these "reverts" are on core functions because of this fact.

Mamaduka replied:
In this case, it should never contain HTML since the value is for the `datetime` attribute.

P.S. There's similar escape login in the latest-comments block.

getdave replied:
> it should never contain HTML since the value is for the datetime attribute.

That's the key point. Thanks 👍

getdave replied:
✅

getdave commented on packages/block-library/src/latest-comments/index.php:
```diff
@@ -86,14 +86,14 @@ function render_block_core_latest_comments( $attributes = array() ) {
 			$author_markup = '';
 			if ( $author_url ) {
-				$author_markup .= '<a class="wp-block-latest-comments__comment-author" href="' . esc_url( $author_url ) . '">' . esc_html( get_comment_author( $comment ) ) . '</a>';
+				$author_markup .= '<a class="wp-block-latest-comments__comment-author" href="' . esc_url( $author_url ) . '">' . get_comment_author( $comment ) . '</a>';
```
Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/latest-comments/index.php:
```diff
@@ -86,14 +86,14 @@ function render_block_core_latest_comments( $attributes = array() ) {
+				$author_markup .= '<a class="wp-block-latest-comments__comment-author" href="' . esc_url( $author_url ) . '">' . get_comment_author( $comment ) . '</a>';
 			} else {
-				$author_markup .= '<span class="wp-block-latest-comments__comment-author">' . esc_html( get_comment_author( $comment ) ) . '</span>';
+				$author_markup .= '<span class="wp-block-latest-comments__comment-author">' . get_comment_author( $comment ) . '</span>';
```
Core function - can be trusted, maybe filtered.

Mamaduka replied:
LGTM 👍

getdave commented on packages/block-library/src/latest-comments/index.php:
```diff
@@ -86,14 +86,14 @@ function render_block_core_latest_comments( $attributes = array() ) {
 			// `_draft_or_post_title` calls `esc_html()` so we don't need to wrap that call in
 			// `esc_html`.
-			$post_title = '<a class="wp-block-latest-comments__comment-link" href="' . esc_url( get_comment_link( $comment ) ) . '">' . esc_html( wp_latest_comments_draft_or_post_title( $comment->comment_post_ID ) ) . '</a>';
+			$post_title = '<a class="wp-block-latest-comments__comment-link" href="' . get_comment_link( $comment ) . '">' . wp_latest_comments_draft_or_post_title( $comment->comment_post_ID ) . '</a>';
```
Core function - can be trusted, maybe filtered.

Mamaduka replied:
Core escapes `get_comment_link` - https://github.com/WordPress/wordpress-develop/blob/f0180f59ea00683647c602941c1f2b422192df46/src/wp-includes/comment.php#L3675

getdave replied:
✅

getdave committed:
Fix incorrect removal of date function. Remove escaping.

getdave commented on packages/block-library/src/latest-comments/index.php:
```diff
@@ -106,7 +106,7 @@ function render_block_core_latest_comments( $attributes = array() ) {
 					'<time datetime="%1$s" class="wp-block-latest-comments__comment-date">%2$s</time>',
 					esc_attr( get_comment_date( 'c', $comment ) ),
-					esc_html( date_i18n( get_option( 'date_format' ), get_comment_date( 'U', $comment ) ) )
+					date_i18n( get_option( 'date_format' ), get_comment_date( 'U', $comment ) )
```
Core function - can be trusted, maybe filtered.

getdave replied:
✅

getdave commented on packages/block-library/src/latest-posts/index.php:
```diff
@@ -119,7 +119,7 @@ function render_block_core_latest_posts( $attributes ) {
 				$list_items_markup .= sprintf(
 					'<div class="wp-block-latest-posts__post-author">%1$s</div>',
-					esc_html( $byline )
+					$byline
```
Generated by Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/latest-posts/index.php:
```diff
@@ -128,7 +128,7 @@ function render_block_core_latest_posts( $attributes ) {
 				'<time datetime="%1$s" class="wp-block-latest-posts__post-date">%2$s</time>',
 				esc_attr( get_the_date( 'c', $post ) ),
-				esc_html( get_the_date( '', $post ) )
+				get_the_date( '', $post )
```
Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/page-list/index.php:
```diff
@@ -188,10 +188,10 @@ function block_core_page_list_render_nested_page_list( $open_submenus_on_click,
 
 		if ( isset( $page['children'] ) && $is_navigation_child && $open_submenus_on_click ) {
-			$markup .= '<button aria-label="' . esc_attr( $aria_label ) . '" class="' . esc_attr( $navigation_child_content_class ) . ' wp-block-navigation-submenu__toggle" aria-expanded="false">' . esc_html( $title ) . '<span class="wp-block-page-list__submenu-icon wp-block-navigation__submenu-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" viewBox="0 0 12 12" fill="none" aria-hidden ...
+			$markup .= '<button aria-label="' . esc_attr( $aria_label ) . '" class="' . esc_attr( $navigation_child_content_class ) . ' wp-block-navigation-submenu__toggle" aria-expanded="false">' . $title . '<span class="wp-block-page-list__submenu-icon wp-block-navigation__submenu-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" viewBox="0 0 12 12" fill="none" aria-hidden="true" focu ...
```
Generated by Core function - can be trusted, maybe filtered.

Mamaduka replied:
The value is sanitized above using `wp_kses` 👍

getdave replied:
If anything we should be the `wp_kses` inline.

Mamaduka replied:
The `$title` is used by aria-label, making sense to "escape" once and re-use.

getdave replied:
Agreed.

getdave commented on packages/block-library/src/page-list/index.php:
```diff
@@ -188,10 +188,10 @@ function block_core_page_list_render_nested_page_list( $open_submenus_on_click,
 			'</button>';
 		} else {
-			$markup .= '<a class="wp-block-pages-list__item__link' . esc_attr( $navigation_child_content_class ) . '" href="' . esc_url( $page['link'] ) . '"' . $aria_current . '>' . esc_html( $title ) . '</a>';
+			$markup .= '<a class="wp-block-pages-list__item__link' . esc_attr( $navigation_child_content_class ) . '" href="' . esc_url( $page['link'] ) . '"' . $aria_current . '>' . $title . '</a>';
```
Generated by Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/page-list/index.php:
```diff
@@ -316,8 +316,8 @@ function render_block_core_page_list( $attributes, $content, $block ) {
 		array(
-			'class' => esc_attr( $css_classes ),
-			'style' => esc_attr( $style_attribute ),
+			'class' => $css_classes,
+			'style' => $style_attribute,
```
Not required. Handled by function.

Mamaduka replied:
Nice catch.

getdave commented on packages/block-library/src/post-comments-link/index.php:
```diff
@@ -33,7 +33,7 @@ function render_block_core_post_comments_link( $attributes, $content, $block ) {
 			/* translators: %s post title */
 			__( 'No comments<span class="screen-reader-text"> on %s</span>' ),
-			esc_html( $post_title )
+			$post_title
```
Generated by Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/post-comments-link/index.php:
```diff
@@ -44,7 +44,7 @@ function render_block_core_post_comments_link( $attributes, $content, $block ) {
 			),
 			esc_html( number_format_i18n( $comments_number ) ),
-			esc_html( $post_title )
+			$post_title
```
Generated by Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/post-date/index.php:
```diff
@@ -23,7 +23,7 @@ function render_block_core_post_date( $attributes, $content, $block ) {
 	$formatted_date     = get_the_date( isset( $attributes['format'] ) ? $attributes['format'] : '', $post_ID );
 	if ( isset( $attributes['isLink'] ) && $attributes['isLink'] ) {
-		$formatted_date = sprintf( '<a href="%1s">%2s</a>', get_the_permalink( $post_ID ), esc_html( $formatted_date ) );
+		$formatted_date = sprintf( '<a href="%1s">%2s</a>', get_the_permalink( $post_ID ), $formatted_date );
```
Generated by Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/site-tagline/index.php:
```diff
@@ -23,7 +23,7 @@ function render_block_core_site_tagline( $attributes ) {
 		'<p %1$s>%2$s</p>',
 		$wrapper_attributes,
-		esc_html( $site_tagline )
+		$site_tagline
```
Generated by Core function - can be trusted, maybe filtered.

getdave commented on packages/block-library/src/site-title/index.php:
```diff
@@ -36,7 +36,7 @@ function render_block_core_site_title( $attributes ) {
 			$link_attrs[] = 'target="_blank"';
 		}
-		$site_title = sprintf( '<a %1$s>%2$s</a>', implode( ' ', $link_attrs ), esc_html( $site_title ) );
+		$site_title = sprintf( '<a %1$s>%2$s</a>', implode( ' ', $link_attrs ), $site_title );
```
Generated by Core function - can be trusted, maybe filtered.

Mamaduka replied:
Do you remember if the site title can contain HTML?

I think we should escape here. The `get_bloginfo` calls the `get_option` directly and doesn't do anything else.

getdave replied:
These scenarios are confusing. If the user has chosen to use HTML in their site title should we allow it to be output to the frontend?

Mamaduka replied:
I can see that the "adminbar" escapes the site title.

![CleanShot 2022-02-07 at 16 25 36](https://user-images.githubusercontent.com/240569/152787814-f8128479-5a96-453e-92bf-576e051a140e.png)

getdave replied:
I've restored the $site_title escaping. 

See https://github.com/WordPress/gutenberg/pull/38569/commits/7456e9dce1ac148f1d0e24bcb8694d3ef1fc212f

getdave committed:
Post title block - revert escaping of core function

getdave committed:
Consider rich text in more text output escaping

getdave commented on packages/block-library/src/post-excerpt/index.php:
```diff
@@ -24,7 +24,7 @@ function render_block_core_post_excerpt( $attributes, $content, $block ) {
 	}
 
-	$more_text           = ! empty( $attributes['moreText'] ) ? '<a class="wp-block-post-excerpt__more-link" href="' . esc_url( get_the_permalink( $block->context['postId'] ) ) . '">' . esc_html( $attributes['moreText'] ) . '</a>' : '';
+	$more_text           = ! empty( $attributes['moreText'] ) ? '<a class="wp-block-post-excerpt__more-link" href="' . esc_url( get_the_permalink( $block->context['postId'] ) ) . '">' . wp_kses_post( $attributes['moreText'] ) . '</a>' : '';
```
More text may be rich text and contain some HTML. See https://github.com/WordPress/gutenberg/issues/38560

getdave committed:
Restore esaping of comment link

Addresses https://github.com/WordPress/gutenberg/pull/38569#discussion_r800572147

getdave committed:
Restore escaping of attribute

Addresses https://github.com/WordPress/gutenberg/pull/38569#discussion_r800532165

getdave committed:
Restore escaping of site title

See https://github.com/WordPress/gutenberg/pull/38569#discussion_r800534754

kevin940726 reviewed (approved):
Code looks good, let's just merge it 👍 .
